Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp

```diff
@@ -0,0 +1,333 @@
+    {
+        auto* self = static_cast<BaseSinkPositionFlushWorkaroundProbe*>(userData);
+        GRefPtr<GstBaseSink> basesink = adoptGRef(GST_BASE_SINK(gst_pad_get_parent(pad)));
+        GST_TRACE_OBJECT(pad, "m_isHandlingFlushStop: %s, pad is flushing: %s, received: %" GST_PTR_FORMAT, boolForPrinting(self->m_isHandlingFlushStop), boolForPrinting(GST_PAD_IS_FLUSHING(pad)), info->data);
+        if (!self->m_isHandlingFlushStop && (info->type & GST_PAD_PROBE_TYPE_EVENT_FLUSH) && GST_EVENT_TYPE(GST_PAD_PROBE_INFO_EVENT(info)) == GST_EVENT_FLUSH_STOP
+            && flushIsNonResetting(GST_PAD_PROBE_INFO_EVENT(info)) && basesink->segment.format != GST_FORMAT_UNDEFINED)
+        {
+            GST_DEBUG_OBJECT(pad, "Received non-resetting FLUSH_STOP while we have a valid segment... propagating the FLUSH_STOP.");
+
+            self->m_isHandlingFlushStop = true;
+            gst_pad_send_event(pad, GST_EVENT(info->data));
+            self->m_isHandlingFlushStop = false;
+
+            // For non-resetting FLUSH_STOP events basesink preserves the segment but still sets have_newsegment to FALSE.
+            // Until the fix was introduced, this makes gst_base_sink_get_position() fail.
+            // The workaround consists on setting have_newsegment to TRUE in that case after the FLUSH_STOP is handled
+            // internally by basesink.
+            //
+            // This is a reasonably safe workaround, as the only other place where have_newsegment is read is in a warning
+            // message in gst_base_sink_chain_unlocked() for when a buffer is pushed without a segment, something that
+            // shouldn't happen in the first place.
+            GST_DEBUG_OBJECT(pad, "Non-resetting FLUSH_STOP has been handled by this element and its downstream. Now force position queries to work by setting basesink->have_newsegment to TRUE.");
+            basesink->have_newsegment = TRUE;
+
+            return GST_PAD_PROBE_HANDLED;
+        }
+        return GST_PAD_PROBE_OK;
+    }
+
+    static void* newUserData() { return new BaseSinkPositionFlushWorkaroundProbe; }
+    static void deleteUserData(void* self) { delete static_cast<BaseSinkPositionFlushWorkaroundProbe*>(self); }
+};
+
+bool BaseSinkPositionFlushWorkaroundProbe::s_isNeeded;
+
+// Workaround for https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/2413 appsink: Fix race condition on caps handling
+// Fix landed upstream in 1.20.3, 1.21.1, 1.22.0.
+
+class AppSinkFlushCapsWorkaroundProbe {
+public:
+    static bool isNeeded()
+    {
+        static std::once_flag onceFlag;
+        std::call_once(onceFlag, initializeIsNeeded);
+        return s_isNeeded;
+    }
+
+    static void installIfNeeded(GstAppSink* appsink)
+    {
+        ASSERT(GST_IS_APP_SINK(appsink));
+        if (!isNeeded())
+            return;
+
+        GRefPtr<GstPad> pad = gst_element_get_static_pad(GST_ELEMENT(appsink), "sink");
```

Comment:
missing `adoptGRef()` here

Reply:
TRUE!

Reply:
Good catch. I can make a follow-up patch to fix it.